templates/README.md

```diff
@@ -0,0 +1,221 @@
+# Template development guidelines
+
+The idea is to reuse and enhance the concept of OpenShift templates and its parameter expansion and also to reuse the kubevirt’s VM.spec structure as a DTO that will provide data to the parameter expansion.
+
+Since the templates will contain lots of redundancies, there should be a build time templating mechanism for composing the final templates from snippets.
+
+Also, there might be multiple different OSes, flavors, sizes mentioned by any single template file if they share the same configuration.
+
+## User Experience
+
+The flow should be the same as usual with regards to openshift templates. The user either takes a local template and generates a VM object.
+
+Or the user gets a template from kubernetes and invokes the proper REST command on top of it to keep the VM linked to the template or edit it in accordance to the template.
+
+The UI can operate on top of template parameters and/or the Template.VM.spec DOM directly.
+
+## Example file
+
+### template.yaml
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: Template # Openshift kind of template or something "better"
+metadata:
+  name: windows-10
+  annotations:
+    description: "Basic windows 10 template"
+    
+    # This set of annotations is meant as a hint only, it is
+    # not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded here informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+    defaults.template.cnv.io/disk: default-disk
+    defaults.template.cnv.io/volume: default-volume
+    defaults.template.cnv.io/network: default-network
+
+    # Replacements useful for finishing the transformation
+    # This is an extension we could use for specifying common
+    # fields for all disks/nics. Requires extra implementation
+    # that will do the find and replace using a syntax documented
+    # here: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+    template.cnv.io/replace:
```

Comment:
… and tehse and other metadata informations should be part of the Template or fo the VM.

A benefit of having it on the VM is that the UI only needs to look at VMs for editing it.
And does not need to reconcile with the original Temaplte.

Also: A template can contain more than one entity, this needs to be reflected in the json paths .. ias it could be possibe that we also include a service, whichmight need validation ifnromations as well.

Reply:
This is just a hint for the UI to know how a disk / nic /.. should look like. We can do without it if we can provide the necessary details like default disk type (scsi, virtio) using some other VM.spec attribute.